GVFS/GVFS.Common/Maintenance/PackfileMaintenanceStep.cs

```diff
@@ -0,0 +1,122 @@
+                                                          .ToList();
+
+            foreach (DirectoryItemInfo info in packDirContents)
+            {
+                if (string.Equals(Path.GetExtension(info.Name), ".idx", StringComparison.OrdinalIgnoreCase))
+                {
+                    string pairedPack = info.FullName.Substring(0, info.FullName.Length - "idx".Length) + "pack";
+
+                    if (!this.Context.FileSystem.FileExists(pairedPack))
+                    {
+                        if (this.Context.FileSystem.TryDeleteFile(info.FullName))
+                        {
+                            yield return info.Name;
+                        }
+                    }
+                }
+            }
+        }
+
+        protected override void PerformMaintenance()
+        {
+            using (ITracer activity = this.Context.Tracer.StartActivity(this.Area, EventLevel.Informational, Keywords.Telemetry, metadata: null))
+            {
+                // forceRun is only currently true for functional tests
+                if (!this.forceRun)
+                {
+                    if (!this.EnoughTimeBetweenRuns())
+                    {
+                        activity.RelatedWarning($"Skipping {nameof(LooseObjectsStep)} due to not enough time between runs");
+                        return;
+                    }
+
+                    IEnumerable<int> processIds = this.RunningGitProcessIds();
+                    if (processIds.Any())
+                    {
+                        activity.RelatedWarning($"Skipping {nameof(LooseObjectsStep)} due to git pids {string.Join(",", processIds)}", Keywords.Telemetry);
+                        return;
+                    }
+                }
+
+                this.CountPackFiles(out int beforeCount, out long beforeSize);
+                GitProcess.Result result = this.RunGitCommand((process) => process.MultiPackIndexExpire(this.Context.Enlistment.GitObjectsRoot));
+                List<string> staleIdxFiles = this.CleanStaleIdxFiles().ToList();
+                this.CountPackFiles(out int expireCount, out long expireSize);
+                GitProcess.Result result2 = this.RunGitCommand((process) => process.MultiPackIndexRepack(this.Context.Enlistment.GitObjectsRoot, this.batchSize));
+                this.CountPackFiles(out int afterCount, out long afterSize);
+
+                EventMetadata metadata = new EventMetadata();
+                metadata.Add("GitObjectsRoot", this.Context.Enlistment.GitObjectsRoot);
+                metadata.Add("BatchSize", this.batchSize);
+                metadata.Add(nameof(beforeCount), beforeCount);
+                metadata.Add(nameof(beforeSize), beforeSize);
+                metadata.Add(nameof(expireCount), expireCount);
+                metadata.Add(nameof(expireSize), expireSize);
+                metadata.Add(nameof(afterCount), afterCount);

```

Comment:
Out of curiosity will the repack save any us any space.  I understand it will reduce the number of files just curious is there are any other gains.

Reply:
It will save space because we will not blindly copy every object out of the packs, but only those referred to by the multi-pack-index. After our first 'expire' command, there are usually a lot of packs that have less than 50 referenced objects (but actually have thousands of duplicate objects).

While we have a 2 GB batch size, we frequently will write a pack that is 200 MB instead.